Branch-SDK/src/main/java/io/branch/referral/BillingGooglePlay.kt

```diff
@@ -0,0 +1,243 @@
+                product.subscriptionOfferDetails?.get(0)?.pricingPhases?.pricingPhaseList?.get(0)
+
+            val currency = pricingPhaseList?.let {
+                CurrencyType.valueOf(
+                    it.priceCurrencyCode
+                )
+            }
+
+            val price = ((pricingPhaseList?.priceAmountMicros ?: 0) / 1000000).toDouble()
+
+            return BranchUniversalObject()
+                .setCanonicalIdentifier(product.productId)
+                .setTitle(product.title)
+                .setContentMetadata(
+                    ContentMetadata()
+                        .addCustomMetadata("product_type", product.productType)
+                        .setPrice(price, currency)
+                        .setProductName(product.name)
+                        .setQuantity(1.0)
+                        .setContentSchema(BranchContentSchema.COMMERCE_PRODUCT)
+
+                )
+        } else {
+            return BranchUniversalObject()
+        }
+    }
+
+
+    private fun createBUOWithInAppProductDetails(
+        product: ProductDetails?,
+        quantity: Int
+    ): BranchUniversalObject {
+        if (product != null) {
+
+            val currency =
+                CurrencyType.valueOf(product.oneTimePurchaseOfferDetails!!.priceCurrencyCode)
+            val price =
+                (product.oneTimePurchaseOfferDetails!!.priceAmountMicros / 1000000).toDouble()
+
+            return BranchUniversalObject()
+                .setCanonicalIdentifier(product.productId)
+                .setTitle(product.title)
+                .setContentMetadata(
+                    ContentMetadata()
+                        .addCustomMetadata("product_type", product.productType)
+                        .setPrice(price, currency)
+                        .setProductName(product.name)
+                        .setQuantity(quantity.toDouble())
+                        .setContentSchema(BranchContentSchema.COMMERCE_PRODUCT)
+                )
+        } else {
+            return BranchUniversalObject()
+        }
+    }
+
+    private fun createAndLongEventForPurchase(
+        context: Context,
+        purchase: Purchase,
+        contentItems: List<BranchUniversalObject>,
+        currency: CurrencyType,
+        revenue: Double,
+        productType: String
+    ) {
+        BranchEvent(BRANCH_STANDARD_EVENT.PURCHASE)
+            .setCurrency(currency)
+            .setDescription(purchase.orderId)
+            .setCustomerEventAlias(productType)
+            .setRevenue(revenue)
+            .addCustomDataProperty("package_name", purchase.packageName)
+            .addCustomDataProperty("order_id", purchase.orderId)
+            .addCustomDataProperty("logged_from_IAP", "true")
+            .addCustomDataProperty("is_auto_renewing", purchase.isAutoRenewing.toString())
+            .addCustomDataProperty("purchase_token", purchase.purchaseToken)
+            .addContentItems(contentItems)
+            .logEvent(context)
```

Comment:
Is there a case when we'd make 10 consecutive network requests and need them all to complete before some other action? (like updating a ui?)